tinygrad/features/multi.py

```diff
@@ -1,16 +1,38 @@
 from __future__ import annotations
 from typing import Optional, Union, Any, Tuple, List
 import functools, itertools, operator
-from tinygrad.helpers import all_same, dedup, round_up, prod, DEBUG
+from tinygrad.helpers import all_same, all_int, dedup, round_up, prod, DEBUG
 from tinygrad.dtype import DType, Scalar
 from tinygrad.ops import BinaryOps, LoadOps, UnaryOps, TernaryOps, ReduceOps
 from tinygrad.lazy import LazyBuffer
 from tinygrad.shape.shapetracker import sint
 
-def all_reduce(op:ReduceOps, lbs):
-  # TODO: replace this with ring reduce
+def ring_allreduce(op: ReduceOps, lbs: List[LazyBuffer]):
+  assert all_int(lbs[0].shape), f"does not support symbolic shape {lbs[0].shape}"
+  assert all_same([lb.shape[0] for lb in lbs]), "allreduce with uneven shards is undefined"
```

Comment:
why is this not defined?

Reply:
![image](https://github.com/tinygrad/tinygrad/assets/83587632/995473c8-9835-4461-b78c-e82c57ec9e82)
Because elementwise adds with different shapes are undefined? I thought about just padding with zeroes but because + op in tinygrad doesn't do this i decided that i shouldn't do either.